packages/playwright-core/src/cli/docker.ts

```diff
@@ -0,0 +1,380 @@
+}
+
+async function findRunningDockerContainerId(): Promise<string|undefined> {
+  const containers: (Container[]|undefined) = await getJSON('/containers/json');
+  if (!containers)
+    return undefined;
+  const dockerImage = await findDockerImage(VRT_IMAGE_NAME);
+  const container = dockerImage ? containers.find((container: Container) => container.ImageID === dockerImage.Id) : undefined;
+  return container?.State === 'running' ? container.Id : undefined;
+}
+
+interface ContainerOptions {
+  image: DockerImage;
+  autoRemove: boolean;
+  command?: string[];
+  ports?: Number[];
+  name?: string;
+}
+
+async function launchContainer(options: ContainerOptions): Promise<string> {
+  const ExposedPorts: any = {};
+  const PortBindings: any = {};
+  for (const port of (options.ports ?? [])) {
+    ExposedPorts[`${port}/tcp`] = {};
+    PortBindings[`${port}/tcp`] = [{ HostPort: port + '' }];
+  }
+  const container = await postJSON(`/containers/create` + (options.name ? '?name=' + options.name : ''), {
+    Cmd: options.command,
+    AttachStdout: true,
+    AttachStderr: true,
+    Image: options.image.Id,
+    ExposedPorts,
+    HostConfig: {
+      Init: true,
+      AutoRemove: options.autoRemove,
+      ShmSize: 2 * 1024 * 1024 * 1024,
+      PortBindings,
+    },
+  });
+  await postJSON(`/containers/${container.Id}/start`);
+  return container.Id;
+}
+
+async function getJSON(url: string): Promise<any> {
+  const result = await callDockerAPI('get', url);
+  if (!result)
+    return result;
+  return JSON.parse(result);
+}
+
+async function postJSON(url: string, json: any = undefined) {
+  const result = await callDockerAPI('post', url, json ? JSON.stringify(json) : undefined);
+  if (!result)
+    return result;
+  return JSON.parse(result);
+}
+
+const DOCKER_API_VERSION = '1.41';
+
+function callDockerAPI(method: 'post'|'get'|'delete', url: string, body: Buffer|string|undefined = undefined): Promise<string> {
+  const dockerSocket = process.platform === 'win32' ? '\\\\.\\pipe\\docker_engine' : '/var/run/docker.sock';
+  return new Promise((resolve, reject) => {
+    const request = http.request({
+      socketPath: dockerSocket,
+      path: `/v${DOCKER_API_VERSION}${url}`,
+      method,
+    }, (response: http.IncomingMessage) => {
+      let body = '';
+      response.on('data', function(chunk){
+        body += chunk;
+      });
+      response.on('end', function(){
+        if (!response.statusCode || response.statusCode < 200 || response.statusCode >= 300)
+          reject(new Error(`${method} ${url} FAILED with statusCode ${response.statusCode} and body\n${body}`));
+        else
+          resolve(body);
+      });
+    });
+    request.on('error', function(e){
+      reject(e);
+    });
+    if (body) {
+      request.setHeader('Content-Type', 'application/json');
+      request.setHeader('Content-Length', body.length);
+      request.write(body);
+    } else {
+      request.setHeader('Content-Type', 'text/plain');
```

Comment:
why set the content type if there is no body?

Reply:
This is to get plain response